com.unity.sg2/Editor/GraphUI/Importers/ShaderGraphAsset.cs

```diff
@@ -7,108 +8,146 @@
-            var graph = asset.ResolveGraph();
-            model.Init(graph);
+            if (!asset.ShaderGraphModel.IsSubGraph)
+            {
+                // build shader and setup supplementary assets
+                var key = Registry.ResolveKey<ShaderGraphContext>();
+                var node = asset.ShaderGraphModel.GraphHandler.GetNode(key.Name);
+                string shaderCode = Interpreter.GetShaderForNode(node, asset.ShaderGraphModel.GraphHandler, asset.ShaderGraphModel.GraphHandler.registry, out var defaultTextures);
+                var shader = ShaderUtil.CreateShaderAsset(ctx, shaderCode, false);
+                Material mat = new (shader);
+                foreach (var def in defaultTextures)
+                {
+                    mat.SetTexture(def.Item1, def.Item2);
+                }
+                Texture2D texture = Resources.Load<Texture2D>("Icons/sg_graph_icon");
 
-            // build shader and setup supplementary assets
-            var reg = ShaderGraphRegistryBuilder.CreateDefaultRegistry();
-            var key = Registry.ResolveKey<ShaderGraphContext>();
-            var node = model.ShaderGraphModel.GraphHandler.GetNodeReader(key.Name);
-            string shaderCode = Interpreter.GetShaderForNode(node, graph, reg, out var defaultTextures);
-            var shader = ShaderUtil.CreateShaderAsset(ctx, shaderCode, false);
-            Material mat = new (shader);
-            foreach (var def in defaultTextures)
+                ctx.AddObjectToAsset("Shader", shader, texture);
+                ctx.SetMainObject(shader);
+                ctx.AddObjectToAsset("Material", mat);
+                ctx.AddObjectToAsset("Asset", asset);
+            }
+            else // is subgraph
             {
-                mat.SetTexture(def.Item1, def.Item2);
+                Texture2D texture = Resources.Load<Texture2D>("Icons/sg_subgraph_icon");
+
+                ctx.AddObjectToAsset("AssetHelper", asset, texture);
+                ctx.SetMainObject(asset);
             }
-            Texture2D texture = Resources.Load<Texture2D>("Icons/sg_graph_icon");
+        }
+    }
+
+
+
 
-            ctx.AddObjectToAsset("MainAsset", shader, texture);
-            ctx.SetMainObject(shader);
-            ctx.AddObjectToAsset("View", model);
-            ctx.AddObjectToAsset("Material", mat);
-            ctx.AddObjectToAsset("AssetHelper", asset); // so we can resolve GraphHandler in OnEnable
+    [Serializable]
+    internal class SerializableGraphHandler : ISerializationCallbackReceiver
+    {
+        [SerializeField]
+        string json = "";
+
+        [NonSerialized]
+        GraphHandler m_graph;
+
+        // Provide a previously initialized graphHandler-- round-trip it for ownership.
+        public void Init(GraphHandler value)
+        {
+            json = value.ToSerializedFormat();
+            var reg = ShaderGraphRegistryBuilder.CreateDefaultRegistry(); // TODO: Singleton?
```

Comment:
This TODOs been addressed no? the registry is a static singleton now

Reply:
It's a fake solution right now. I _had_ to do it to get correct initialization behavior, but it breaks lifecycle assumptions the the UI has, so singleton~ing it prevents it from being able to fetch and cache the UI Descriptors.

Reply:
This has been addressed.